internal/attestation/crafter/materials/materials.go

```diff
@@ -43,43 +44,56 @@ type crafterCommon struct {
+		return nil, fmt.Errorf("this file is too big for the %s CAS backend, please contact your administrator: fileSize=%s, maxSize=%s", backend.Name, bytefmt.ByteSize(uint64(len(artifactData))), bytefmt.ByteSize(uint64(backend.MaxSize)))
+	}
+	return uploadAndCraft(ctx, input, backend, bytes.NewBuffer(artifactData), filename, h.String(), l)
+}
+
+// uploadAndCraftFromFile uploads the artifact to CAS and crafts the material
 // this function is used by all the uploadable artifacts crafters (SBOMs, JUnit, and more in the future)
-func uploadAndCraft(ctx context.Context, input *schemaapi.CraftingSchema_Material, backend *casclient.CASBackend, artifactPath string, l *zerolog.Logger) (*api.Attestation_Material, error) {
-	// 1 - Check the file can be stored in the provided CAS backend
+func uploadAndCraftFromFile(ctx context.Context, input *schemaapi.CraftingSchema_Material, backend *casclient.CASBackend, artifactPath string, l *zerolog.Logger) (*api.Attestation_Material, error) {
 	result, err := fileStats(artifactPath)
 	if err != nil {
 		return nil, fmt.Errorf("getting file stats: %w", err)
 	}
 	defer result.r.Close()
 
-	l.Debug().Str("filename", result.filename).Str("digest", result.digest).Str("path", artifactPath).
-		Str("size", bytefmt.ByteSize(uint64(result.size))).
-		Str("max_size", bytefmt.ByteSize(uint64(backend.MaxSize))).
-		Str("backend", backend.Name).Msg("crafting file")
-
 	// If there is a max size set and the file is bigger than that, return an error
 	if backend.MaxSize > 0 && result.size > backend.MaxSize {
 		return nil, fmt.Errorf("this file is too big for the %s CAS backend, please contact your administrator: fileSize=%s, maxSize=%s", backend.Name, bytefmt.ByteSize(uint64(result.size)), bytefmt.ByteSize(uint64(backend.MaxSize)))
 	}
 
+	return uploadAndCraft(ctx, input, backend, result.r, result.filename, result.digest, l)
+}
+
+func uploadAndCraft(ctx context.Context, input *schemaapi.CraftingSchema_Material, backend *casclient.CASBackend, reader io.Reader, filename, hash string, l *zerolog.Logger) (*api.Attestation_Material, error) {
+	buf := bufio.NewReader(reader)
+
+	l.Debug().Str("filename", filename).Str("digest", hash).
+		Str("size", bytefmt.ByteSize(uint64(buf.Size()))).
+		Str("max_size", bytefmt.ByteSize(uint64(backend.MaxSize))).
+		Str("backend", backend.Name).Msg("crafting file")
+
 	material := &api.Attestation_Material{
 		MaterialType: input.Type,
 		M: &api.Attestation_Material_Artifact_{
 			Artifact: &api.Attestation_Material_Artifact{
 				Id:        input.Name,
-				Name:      result.filename,
-				Digest:    result.digest,
+				Name:      filename,
+				Digest:    hash,
 				IsSubject: input.Output,
 			},
 		},
 	}
 
-	// 2 - Upload the file to CAS
+	// Upload the file to CAS
 	if backend.Uploader != nil {
 		l.Debug().Str("backend", backend.Name).Msg("uploading")
 
-		_, err = backend.Uploader.UploadFile(ctx, artifactPath)
+		_, err := backend.Uploader.Upload(ctx, buf, hash, filename)
```

Comment:
I am not 100% about these changes. Especially the fact that the content and the ahs are now separate. I am wondering if an alternative less optimal, but that will require less changes would have been to store the content in a temporary file. Not sure, I'd love to get your take

Reply:
Taking into account that the grunt of these changes are because of we want to provide some modified versions of a provided file, an alternative, would have been to during the attestation material craft, take the resulted marshalled data and store it in a temporary file and use that one instead.

Then, the rest of the code wouldn't have changed.

Reply:
Yes, cleaner for sure. Let me refactor a little bit the change.